File: examples/bionetProxyClient.py
#!/usr/bin/python3

"""
to run, first start bionetProxy:
node bionetProxyServer -protocol=https -host=endylab.stanford.edu -listen=8088 -username=xxxx -password=xxxx
"""

import asyncio
import websockets
from io import StringIO
import json

def oolog(msg):
    print(msg)

class BionetApi():

    ws_open = False
    eventLoop = None

    def __init__(self, socketUrl):
        self.wsUrl = socketUrl

    async def arpc(self, request, consumer):
        io = StringIO()
        try:
            json.dump(request, io)
        except:
            oolog("arpc: error in json.dump")

        await self.websocket.send(io.getvalue())
        message = await self.websocket.recv()

        try:
            jsonMessage = json.loads(message)
            await consumer(jsonMessage)
        except:
            oolog("arpc: error in json.loads"+message)

    def rpc(self, request, consumer):
        async def runRpc():
            async with websockets.connect(
                    self.wsUrl, subprotocols=["bionet-protocol"]) as websocket:
                self.websocket = websocket
                await self.arpc(request, consumer)

        asyncio.get_event_loop().run_until_complete(runRpc())


if __name__ == "__main__":

    async def rpcSearchResultCSV(response):
        oolog('\nrpcSearchResultCSV:\n')

        def writeSearchResultRow(row):
            result = row['value']
            created = result['created']
            updated = result['updated']
            rowOutput = '"{name}","{description}","{sequence}","{created}","{updated}","{id}"'.format(name=result['name'],description=result['description'],sequence=result['sequence'],created=created['user'],updated=updated['user'],id=result['id'])
            oolog(rowOutput)

        for row in response:
            writeSearchResultRow(row)

    async def rpcGetResult(response):
        oolog('\nrpcGetResult:\n')
        io = StringIO()
        json.dump(response, io)
        oolog(io.getvalue())

    bionetApi = BionetApi("ws://localhost:8088")
    bionetApi.rpc(
        {"method": "searchVirtuals", "args": ['r']},
        rpcSearchResultCSV
    )
    bionetApi.rpc(
        {"method": "get", "args": ['v-2773e301-03bd-4599-b46c-6fb46aa4b054']},
        rpcGetResult
    )

File: examples/bionetProxyClient.js
#!/usr/bin/env node

/*
to run, first start bionetProxy:
../src/bionetProxyServer -protocol=https -host=endylab.stanford.edu -listen=8088 -username=xxxx -password=xxxx
*/

const WebSocketClient = require('websocket').client;
const fs = require('fs');

var proto = "ws://"
var host = "localhost"
var port = "8088"
var authToken = null
var hostConfig = null
var proxyHost = null
var proxyProtocol = null
var rpc_method = null
var rpc_args = null
var rpcBatch = null
var username = null
var password = null
var import_file = null
var export_file = null
var export_format = 'json'
var exitWhenDone=false

function getCommandLineArgs() {

    function getParam(value) {
        var data = null
        var pa = value.split('=')
        if (pa.length > 1) {
            data = pa[1]
        }
        return data
    }

    process.argv.forEach(function (val, index, _array) {

        // todo: print params

        if (val.indexOf('-hostConfig') >= 0) {
            hostConfigFile = getParam(val)
            const data = fs.readFileSync(hostConfigFile)
            if (data) {
                hostConfig = JSON.parse(data)
                proxyProtocol = hostConfig.protocol
                proxyHost = hostConfig.host
                authToken = hostConfig.token
            }
        }
        else if (val.indexOf('-protocol') >= 0) {
            proxyProtocol = getParam(val)
        }
        else if (val.indexOf('-host') >= 0) {
            proxyHost = getParam(val)
        }
        else if (val.indexOf('-username') >= 0) {
            username = getParam(val)
        }
        else if (val.indexOf('-password') >= 0) {
            password = getParam(val)
        }
        else if (val.indexOf('-rpc_method') >= 0) {
            rpc_method = getParam(val)
        }
        else if (val.indexOf('-batch') >= 0) {
            var batchFile = getParam(val)
            const data = fs.readFileSync(batchFile)
            if (data) {
                rpcBatch = JSON.parse(data)
            }
        }
        else if (val.indexOf('-import=') >= 0) {
            // todo: generate rpcBatch args from csv
            import_file = getParam(val)
        }
        else if (val.indexOf('-export=') >= 0) {
            // todo: setup export to file
            export_file = getParam(val)
        }
        else if (val.indexOf('-export_format') >= 0) {
            export_format = getParam(val)
        }
        else if (val.indexOf('-rpc_args') >= 0) {
            var rpc_argv = getParam(val)
            rpc_argv = rpc_argv.replace(/`/g, '');
            //todo catch json parse errors
            rpc_args = JSON.parse(rpc_argv)
        }
    });

    // post processing import_file
    if (import_file) {
        const data = fs.readFileSync(import_file)
        //todo: convert data to rpcBatch
    }
}

getCommandLineArgs()

var client = new WebSocketClient();

client.on('connectFailed', function (error) {
    console.log('Connect Error: ' + error.toString());
});

function sendRpcRequest(rpcRequest, _connection) {
    if (_connection.connected) {
        _connection.sendUTF(JSON.stringify(rpcRequest));
    } else {
        console.log('sendRpcRequest: not connected to proxy host, exiting...')
        process.exit()
    }
}

function run(_connection) {
    if (rpcBatch) {
        rpcBatch.forEach(function (rpc_request) {
            sendRpcRequest(rpc_request, _connection);
        })
    } else {
        var rpcRequest = {
            "method": rpc_method,
            "args": rpc_args
        }
        exitWhenDone=true
        sendRpcRequest(rpcRequest, _connection);
    }
}

client.on('connect', function (connection) {
    console.log('bionet proxy webSocket connected');
    connection.on('error', function (error) {
        console.log("Connection Error: " + error.toString());
        process.exit()
    });
    connection.on('close', function () {
        console.log('bionet proxy connection closed');
        process.exit()
    });
    connection.on('message', function (message) {
        if (message.type === 'utf8') {
            jsonMsg = JSON.parse(message.utf8Data)
            console.log(JSON.stringify(jsonMsg, null, 2));
        }
        if (exitWhenDone) process.exit()
    });
    run(connection)
});

var wsUrl = proto + host + ":" + port + "/"
client.connect(wsUrl, 'bionet-protocol');
